plugins/datastreamer_plugin/fastdds/Participant.cpp

```diff
@@ -0,0 +1,456 @@
+    }
+
+    // The rest of maps and variables are destroyed by themselves
+}
+
+////////////////////////////////////////////////////
+// INTERACTION METHODS
+////////////////////////////////////////////////////
+
+// TODO: check if error handling by bool or exception
+bool Participant::register_type_from_xml(
+        const std::string& xml_path)
+{
+    eprosima::fastrtps::xmlparser::XMLP_ret ret =
+            eprosima::fastrtps::xmlparser::XMLProfileManager::loadXMLFile(xml_path);
+    if (eprosima::fastrtps::xmlparser::XMLP_ret::XML_OK != ret)
+    {
+        logWarning(PLOTJUGGLER_FASTDDS, "Error loading XML file: " << xml_path);
+        return false;
+    }
+
+    // Loading an xml does not retrieve the types loaded.
+    // Thus, it is necessary to loop over all types discovered and check if they are registered already
+    refresh_types_registered_();
+
+    return true;
+}
+
+void Participant::create_subscription(
+        const std::string& topic_name)
+{
+    // TODO: check if mutex required
+    DEBUG("Creating subscription for topic: " << topic_name);
+
+    // Check datareader does not exist yet
+    if (readers_.find(topic_name) != readers_.end())
+    {
+        logWarning(PLOTJUGGLER_FASTDDS, "Datareader already exists for topic: " << topic_name);
+        return;
+    }
+
+    // Check the Topic exist, so the type name is known
+    auto topic_type = discovery_database_->find(topic_name);
+    if (topic_type == discovery_database_->end())
+    {
+        logWarning(
+            PLOTJUGGLER_FASTDDS,
+            "Topic " << topic_name << " has not been discovered not exist, so type unknowon");
+        return;
+    }
+
+    // Get TypeName associated with topic name
+    std::string type_name = discovery_database_->operator [](topic_name).first;
+
+    // Create topic
+    eprosima::fastdds::dds::Topic* topic = participant_->create_topic(
+        topic_name,
+        type_name,
+        default_topic_qos_());
+
+    // Create datareader
+    eprosima::fastdds::dds::DataReader* datareader = subscriber_->create_datareader(
+        topic,
+        default_datareader_qos_(),
+        this);     // Mask not required
+
+    // Get Dyn Type for type
+    // This could not fail, as we know type is registered
+    eprosima::fastrtps::types::DynamicType_ptr dyn_type = get_type_registered_(type_name);
+
+    // In case type is not registered, do not create subscription and show error
+    if (!dyn_type)
+    {
+        logError(
+            PLOTJUGGLER_FASTDDS,
+            "Type " << type_name << " is not registered, so datareader cannot be created.");
+        return;
+    }
+
+    // Create Reader Handler with all this information and add it to readers
+    // Create it with specific deleter for reader and topic
+    ReaderHandlerReference new_reader(
+        new ReaderHandler(
+            topic,
+            datareader,
+            dyn_type,
+            listener_),
+        (*reader_handler_deleter_)
```

Comment:
As said, don't wrap a deleter within a unique pointer (very strange an unneeded), nor put it as a class member. Instead just create the deleter here:
```
ReaderHandlerReference new_reader(
        new ReaderHandler(
            topic,
            datareader,
            dyn_type,
            listener_),
        reader_handler_deleter_(participant_, subscriber_)
        );
```

Reply:
It is perfectly ok to have a deleter for each reader, it is very cheap to build and here makes perfect sense, since it provides better encapsulation (apart from simplicity - no need for a global deleter, nor a unique_ptr over it, which is strange)

I fix my proposal, the following should work, and does not require a common `reader_handler_deleter_`
```
ReaderHandlerReference new_reader(
        new ReaderHandler(
            topic,
            datareader,
            dyn_type,
            listener_),
         ReaderHandlerDeleter(participant_, subscriber_)
        );
```

(See https://en.cppreference.com/w/cpp/memory/unique_ptr/unique_ptr for reference)